### ISSUE TITLE
question: custom file transforms

### ISSUE THREAD

**Issue by lifeart** [NullVoxPopuli/ember-addon-migrator#20]
recreated from: https://github.com/NullVoxPopuli/ember-addon-migrator/issues/17#issuecomment-1045339291

Tldr: It will be good to have ability to write per-file transforms for addon migration. Also, it may be useful for future migration cases (to gts/gjs)

Looks like we could thing about foundation to support all this cases in atomic, maintainable way.

For migration case - 

Because one file (eslint config - for example) may be copied into 2+ places (docs, tests, src), likely we should have different transform logic for this places, and it leads us to:

`src/transforms/{targetFilePath__targetFileName}` with 1 export: 

```js
{ 
  async transform(content, meta = {}): string;
}
```

### Example:

Original:

```
.eslintrc.js
```

Moved, but not modified things:

```
src/.eslintrc.js
tests/.eslintrc.js
```

Transforms for this files:

```
src/transforms/src__.eslintrc.js.js
src/transforms/tests__.eslintrc.js.js
```

Every transform could be responsible for only one file, does not have fs access, and receive content as param (for better testing)

With this approach we could create "transform" engine, with ability to maintain multiple files and get atomic community contributions without merge conflicts and with minimal scope (1 script = 1 file to transform)

Helper functions may be located in (tbd)

### Generic transform engine may look like:

```js
async doCopyStaff();
const filePaths = walkSync(wokrspaceDir);

for (const file of filePaths) {
   const transformPath = join(__dirname, file.split('/').join('__'), '.js');
   if (!fs.existsSync(transformPath)) {
      continue;
   }
   const { transform } = await import(transformPath);
   const content = fs.readFileSync(path.join(workspace, file), 'utf8');
   const result = await transform(content, { filePath: file, addon });
   fs.writeFileSync(path.join(workspace, file), result, 'utf8');
}
```

### Things to figure out: 

1. Should we support multiple transform files for one file?
1. How we could support ts/js source files transform? (should it be one transform file, or 2 with 1 reexport)
1. Better file structure for transform files? 
 - in general, we could have every transform in it's own folder, with included tests for it, to get per-transform maintain scope as small as possible.

Example, inspired by: https://github.com/babel-utils/babel-plugin-tester#fixtures

```
transforms/src__eslintrc/index.js
transforms/src__eslintrc/tests/{case_1}/input.json
transforms/src__eslintrc/tests/{case_1}/output.json
```

`transforms/src__eslintrc/index.js`
```js
export const supportedPaths = ['src/.eslintrc.js', 'src/.eslintrc.json'];
export async function transform(content, meta) { }
```

`transforms/src__eslintrc/{case_1}/input.json`
```json
{
 "content": "",
 "meta": { "filePath": "src/.eslintrc.js" }
}
```
